### ISSUE TITLE
yara-python does not "satisfy requirements"

### ISSUE THREAD

**Issue by cryptax** [rednaga/APKiD#7]
Hi,
I've compiled yara-python (3.4.0.999), but APKiD is still complaining it can't find the right version. 

```
[~/softs/APKiD]$ pip install apkid
Downloading/unpacking apkid
  Downloading apkid-0.9.3-py2.py3-none-any.whl
Downloading/unpacking yara-python==3.4.0.999 (from apkid)
  Could not find a version that satisfies the requirement yara-python==3.4.0.999 (from apkid) (from versions: 3.5.0, 3.4.0.00)
Cleaning up...
No distributions matching the version for yara-python==3.4.0.999 (from apkid)
Storing debug log for failure in /tmp/tmpVjlWNk
```

To my understanding, the debug logs show we're trying to download yara-python from the official place (https://pypi.python.org/simple/yara-python/) where there is only a 3.4.0 and a 3.5.0. Not a 3.4.0.999.
This is probably the issue. How do I tell it to use the version I just compiled?!

```
Downloading/unpacking apkid
  Getting page https://pypi.python.org/simple/apkid/
  URLs to search for versions for apkid:
  * https://pypi.python.org/simple/apkid/
  Analyzing links from page https://pypi.python.org/simple/apkid/
    Found link https://pypi.python.org/packages/36/9c/4db1239d0f14835214ec50c696665139a3658e922c3676932cb9094ff2ae/apkid-0.9.2-py2.py3-none-any.whl#md5=3f8cda2b40f18f5252230e781713debc (from https://pypi.python.org/simple/apkid/), version: 0.9.2
    Found link https://pypi.python.org/packages/95/87/919246ba87ea7292c7c01f8d1c80b73631b978320825ed3912a90035a187/apkid-0.9.3-py2.py3-none-any.whl#md5=da56eb6ffb4ab0f8474934dfbc87a7ee (from https://pypi.python.org/simple/apkid/), version: 0.9.3
  Using version 0.9.3 (newest of versions: 0.9.3, 0.9.2)
  Downloading apkid-0.9.3-py2.py3-none-any.whl
  Downloading from URL https://pypi.python.org/packages/95/87/919246ba87ea7292c7c01f8d1c80b73631b978320825ed3912a90035a187/apkid-0.9.3-py2.py3-none-any.whl#md5=da56eb6ffb4ab0f8474934dfbc87a7ee (from https://pypi.python.org/simple/apkid/)
Downloading/unpacking yara-python==3.4.0.999 (from apkid)
  Getting page https://pypi.python.org/simple/yara-python/
  URLs to search for versions for yara-python==3.4.0.999 (from apkid):
  * https://pypi.python.org/simple/yara-python/3.4.0.999
  * https://pypi.python.org/simple/yara-python/
  Getting page https://pypi.python.org/simple/yara-python/3.4.0.999
  Could not fetch URL https://pypi.python.org/simple/yara-python/3.4.0.999: 404 Client Error: Not Found
  Will skip URL https://pypi.python.org/simple/yara-python/3.4.0.999 when looking for download links for yara-python==3.4.0.999 (from apkid)
  Analyzing links from page https://pypi.python.org/simple/yara-python/
    Found link https://pypi.python.org/packages/3e/6c/ec2453fd811ea2f8c157d196b4dd76eb3c229362918fdb661358bd3e8645/yara-python-3.5.0.tar.gz#md5=3184efa3da16e3a8304185f2641c4ffe (from https://pypi.python.org/simple/yara-python/), version: 3.5.0
    Found link https://pypi.python.org/packages/8e/01/b7dce32aa35248e9c640250788cd959245c6f79e6f5d44185ad214efb115/yara-python-3.4.0.00.tar.gz#md5=b28916c20bd8e6a7c4f105e3b496f378 (from https://pypi.python.org/simple/yara-python/), version: 3.4.0.00
  Ignoring link https://pypi.python.org/packages/3e/6c/ec2453fd811ea2f8c157d196b4dd76eb3c229362918fdb661358bd3e8645/yara-python-3.5.0.tar.gz#md5=3184efa3da16e3a
8304185f2641c4ffe (from https://pypi.python.org/simple/yara-python/), version 3.5.0 doesn't match ==3.4.0.999
  Ignoring link https://pypi.python.org/packages/8e/01/b7dce32aa35248e9c640250788cd959245c6f79e6f5d44185ad214efb115/yara-python-3.4.0.00.tar.gz#md5=b28916c20bd8e6a7c4f105e3b496f378 (from https://pypi.python.org/simple/yara-python/), version 3.4.0.00 doesn't match ==3.4.0.999
  Could not find a version that satisfies the requirement yara-python==3.4.0.999 (from apkid) (from versions: 3.5.0, 3.4.0.00)
Cleaning up...
  Removing temporary dir /tmp/pip_build_root...
No distributions matching the version for yara-python==3.4.0.999 (from apkid)
Exception information:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/pip/basecommand.py", line 122, in main
    status = self.run(options, args)
  File "/usr/lib/python2.7/dist-packages/pip/commands/install.py", line 278, in run
    requirement_set.prepare_files(finder, force_root_egg_info=self.bundle, bundle=self.bundle)
  File "/usr/lib/python2.7/dist-packages/pip/req.py", line 1178, in prepare_file
s
    url = finder.find_requirement(req_to_install, upgrade=self.upgrade)
  File "/usr/lib/python2.7/dist-packages/pip/index.py", line 322, in find_requirement
    raise DistributionNotFound('No distributions matching the version for %s' % req)
DistributionNotFound: No distributions matching the version for yara-python==3.4.0.999 (from apkid)
```

This is the version I compiled:
```
$ sudo python setup.py install
[..]
Copying yara_python.egg-info to /usr/local/lib/python2.7/dist-packages/yara_python-3.4.0.999-py2.7.egg-info
```

To summarize what I did exactly:
```bash
$ git clone https://github.com/rednaga/APKiD
$ cd APKiD/
$ git clone https://github.com/rednaga/yara-python
$ cd yara-python/
// fix dex.c //
$ sudo python setup.py install
$ cd ../
$ pip install apkid
```

**Comment by CalebFenton** [comment]
I just tested this on a fresh Linux machine and wasn't able to reproduce the issue. Need to figure out how your environment is different. Could you please paste the output of `sudo python setup.py install` for yara-python and `pip --version` and `python --version` ?

The reason it searches for 3.4.0.999 is so that no official version will be found and there will be an error at install time rather than at run time.

**Comment by cryptax** [comment]
Hi @CalebFenton 

There you go. I have no copying build/lib.linux-x86_64-2.7/yara.so line as you'll notice:
```bash
[~/softs/APKiD/yara-python]
[!689]$ sudo python setup.py install
[sudo] password for axelle: 
running install
running build
running build_ext
running install_lib
running install_egg_info
running egg_info
writing yara_python.egg-info/PKG-INFO
writing top-level names to yara_python.egg-info/top_level.txt
writing dependency_links to yara_python.egg-info/dependency_links.txt
reading manifest file 'yara_python.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'yara_python.egg-info/SOURCES.txt'
removing '/usr/local/lib/python2.7/dist-packages/yara_python-3.4.0.999-py2.7.egg-info' (and everything under it)
Copying yara_python.egg-info to /usr/local/lib/python2.7/dist-packages/yara_python-3.4.0.999-py2.7.egg-info
```

Pip version:
```bash
$ pip --version
pip 1.5.4 from /usr/lib/python2.7/dist-packages (python 2.7)
```

Python version:
```bash
$ python --version
Python 2.7.6
```

**Comment by CalebFenton** [comment]
That line is tricky. It only shows up after you clean and build. I think from your output, it didn't do the build again. I was mostly interested in where it was putting the module, but thought there might be a possibility it was putting it somewhere weird.

Could you try upgrading pip (`pip install --upgrade pip`)? Yours is 26 versions and 2 years behind and I _may_ be relying on some new features.

**Comment by cryptax** [comment]
Ok, I upgraded pip (8.1.2), but I still don't get that line you are looking for, and install of APKid still fails.

```
[!602]$ sudo python setup.py clean
running clean
[!603]$ sudo python setup.py install
running install
running build
running build_ext
running install_lib
running install_egg_info
running egg_info
writing yara_python.egg-info/PKG-INFO
writing top-level names to yara_python.egg-info/top_level.txt
writing dependency_links to yara_python.egg-info/dependency_links.txt
reading manifest file 'yara_python.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'yara_python.egg-info/SOURCES.txt'
removing '/usr/local/lib/python2.7/dist-packages/yara_python-3.4.0.999-py2.7.egg-info' (and everything under it)
Copying yara_python.egg-info to /usr/local/lib/python2.7/dist-packages/yara_python-3.4.0.999-py2.7.egg-info
```

Any hint?

**Comment by CalebFenton** [comment]
Sorry, I left it as a parenthetical but in my tests you have to rm the
build directory.

**Comment by cryptax** [comment]
It re-builds:
```
.linux-x86_64-2.7/yara/libyara/modules/elf.o build/temp.linux-x86_64-2.7/yara/libyara/modules/dex.o build/temp.linux-x86_64-2.7/yara/libyara/modules/hash.o -lcrypto -o build/lib.linux-x86_64-2.7/yara.so
running install_lib
copying build/lib.linux-x86_64-2.7/yara.so -> /usr/local/lib/python2.7/dist-packages
running install_egg_info
running egg_info
writing yara_python.egg-info/PKG-INFO
writing top-level names to yara_python.egg-info/top_level.txt
writing dependency_links to yara_python.egg-info/dependency_links.txt
reading manifest file 'yara_python.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'yara_python.egg-info/SOURCES.txt'
removing '/usr/local/lib/python2.7/dist-packages/yara_python-3.4.0.999-py2.7.egg-info' (and everything under it)
Copying yara_python.egg-info to /usr/local/lib/python2.7/dist-packages/yara_python-3.4.0.999-py2.7.egg-info
```

But still doesn't find it:
```
pip install apkid
Collecting apkid
/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:318: SNIMissingWarning: An HTTPS request has been made, but the SNI (Subject Name Indication) extension to TLS is not available on this platform. This may cause the server to present an incorrect TLS certificate, which can cause validation failures. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#snimissingwarning.
  SNIMissingWarning
/usr/local/lib/python2.7/dist-packages/pip/_vendor/requests/packages/urllib3/util/ssl_.py:122: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
  Using cached apkid-0.9.3-py2.py3-none-any.whl
Collecting yara-python==3.4.0.999 (from apkid)
  Could not find a version that satisfies the requirement yara-python==3.4.0.999 (from apkid) (from versions: 3.4.0.0, 3.5.0)
No matching distribution found for yara-python==3.4.0.999 (from apkid)
```

I'm sorry, I must be a nut!

**Comment by CalebFenton** [comment]
No worries. I want to get this working for you and I'm glad you're taking the time to report your failures!

I'm really not sure what it could be. Python packaging is kind of weird. Looking into similar problems, but a few things you can try:

1. Make sure `pip list` includes `yara-python (3.4.0.999)`.
2. `pip uninstall apkid` and then `pip install apkid --verbose` and post logs here.
3. Nuclear option - `pip uninstall apkid yara-python` (twice just to be sure), delete, re-clone, and re-build everything

**Comment by cryptax** [comment]
Yeah, well I am half pleased ... that it works now (after that proper re-install). Pleased because it works, and Half, because I don't exactly know why ;)
By the way, I still had to patch those compiler issues in yara-python.
Thanks for your help!
